### PR TITLE
Proposal for a more robust implementation.

### PR DESCRIPTION
**Pull request by ju1ius**
Hi,
I spent some type investigating a more robust way to map webpack's path templates to actual output files.

I was hoping to find the perfect plugin event to hook into, but unfortunately most of them are undocumented (maybe because they are internal only).

So I wrote a small parser, that turns a path template string into a PathTemplate object (see `lib/pathTemplate.js` and `test/pathTemplate.test.js`).

This object holds information about the placeholders and static parts of the template string. These informations are used to create a regular expression that can match filenames against the original path template string.

For example the path template string `"[name].[id]_[hash].js[query]"` is internally converted to:

``` js
/^.+?(\.)\d+(_)[0-9A-Fa-f]+(\.js)(?:\?.+?)?$/
```

This allows us to detect sourcemaps and hot updates in a reliable way.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,42 +1,103 @@
...
+            var assetsByChunkName = stats.assetsByChunkName;
+
+            var output = Object.keys(assetsByChunkName).reduce(function (chunkMap, chunkName) {
+                var assets = assetsByChunkName[chunkName];
+                if (!Array.isArray(assets)) {
+                    assets = [assets];
+                }
+                chunkMap[chunkName] = assets.reduce(function (typeMap, asset) {
+                    if (isHMRUpdate(options, asset)) {
+                        return typeMap;
```
Comment (ju1ius) [inline index.js:63]: Here we skip hot module replacement updates

File: index.js [hunk 2]
```diff
@@ -1,42 +1,103 @@
...
-	var _this = this;
-	// var assets = {};
+AssetsWebpackPlugin.prototype = {
+    constructor: AssetsWebpackPlugin,
+
+    getOptions: function (options) {
+        var defaults = {
+            path: '.',
+            filename: 'webpack-assets.json',
+            prettyPrint: false
```
Comment (ju1ius) [inline index.js:21]: Added prettyPrint option for a more human readable output.

File: lib/pathTemplate.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+ *     [, width: maximum hash length for hash & chunkhash placeholders]
+ * }
+ * 
+ * The values in the object conceptually represent a span of literal text followed by a single replacement field.
+ * If there is no literal text (which can happen if two replacement fields occur consecutively),
+ * then prefix will be an empty string.
+ * If there is no replacement field, then the value of placeholder will be null.
+ * If the value of placeholder is either 'hash' or 'chunkhash', then width will be a positive integer.
+ * Otherwise it will be left undefined.
+ */
```
Comment (ju1ius) [inline lib/pathTemplate.js]: Path template string parsing happens here.

File: lib/pathTemplate.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        pos++;
+    }
+
+    return fields;
+}
+
+/**
+ * Returns a RegExp which, given the replacement fields returned by parseTemplate(),
+ * can match a file path against a path template.
+ */
```
Comment (ju1ius) [inline lib/pathTemplate.js]: Here we create the RegExp pattern for matching

File: lib/pathTemplate.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                continue;
+            }
+
+        }
+        prefix += char;
+        pos++;
+    }
+
+    return fields;
+}
```
Comment (ju1ius) [inline lib/pathTemplate.js]: Path template string parsing happens above.

File: lib/pathTemplate.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        }
+        if (i === length - 1) {
+            pattern += '$';
+        }
+
+        return pattern;
+    }, '');
+
+    return new RegExp(pattern);
+}
```
Comment (ju1ius) [inline lib/pathTemplate.js]: RegExp pattern creation for matching happens above

File: lib/pathTemplate.js [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+ * }
+ * 
+ * The values in the object conceptually represent a span of literal text followed by a single replacement field.
+ * If there is no literal text (which can happen if two replacement fields occur consecutively),
+ * then prefix will be an empty string.
+ * If there is no replacement field, then the value of placeholder will be null.
+ * If the value of placeholder is either 'hash' or 'chunkhash', then width will be a positive integer.
+ * Otherwise it will be left undefined.
+ */
+function parseTemplate (str) {
```
Comment (ju1ius) [inline lib/pathTemplate.js:79]: Path template string parsing happens below

File: lib/pathTemplate.js [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    return fields;
+}
+
+/**
+ * Returns a RegExp which, given the replacement fields returned by parseTemplate(),
+ * can match a file path against a path template.
+ */
+function createTemplateMatcher (fields) {
```
Comment (ju1ius) [inline lib/pathTemplate.js:132]: RegExp creation for matching happens below

File: lib/getFileExtension.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+var URL = require('url');
+var path = require('path');
+
+module.exports = function getFileExtension (asset) {
+    var url = URL.parse(asset);
+    var ext = path.extname(url.pathname);
+    return ext ? ext.slice(1) : '';
+}
```
Comment (ju1ius) [inline lib/getFileExtension.js:8]: Use node standard lib :wink:

**Comment by ju1ius** [comment]
As a side note, we could also decide to skip sourcemaps altogether in the output json file.
Since sourcemap-supporting tools detect the sourcemap with the `//# sourceMappingURL=` comment inside the original file, I can't see what's the use case for keeping them.

This would make the plugin code simpler, and the output format even easier to understand.

If you agree with this, I'll update the PR.

File: index.js [hunk 10]
```diff
@@ -1,42 +1,103 @@
...
+                publicPath: true,
+                assets: true,
+                chunks: false,
+                modules: false,
+                source: false,
+                errorDetails: false,
+                timings: false
+            });
+            // publicPath with resolved [hash] placeholder
+            var publicPath = stats.publicPath || '';
```
Comment (ju1ius) [inline index.js:45]: Bonus: we now support [hash] in the public path :wink:

**Comment by sporto** [comment]
Thanks for the PR, You are right the plug-in is getting very complex because of all the flexibility webpack providers. Definitely removing the source maps will simplify it, but I don't really know if anyone is using this.

**Comment by sporto** [comment]
Published 2.3.0

**Pull request merged**